tests/acceptance/25_cf-execd/cf-execd-rpl-functions.c

```diff
@@ -0,0 +1,126 @@
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA
+
+  To the extent this program is licensed as part of the Enterprise
+  versions of CFEngine, the applicable Commercial Open Source License
+  (COSL) may apply to this file if you as a licensee so wish it. See
+  included file COSL.txt.
+*/
+
+/*
+ * A test source file with alternative implementations of some functions. Those
+ * that are affected are generally ifdef'ed out in the original source file by
+ * TEST_CF_EXECD macro.
+ */
+
+#include <platform.h>
+#include <cf-execd-runner.h>
+
+static void *FeedSmtpDirectives(void *data)
+{
+    int socket = *(int *)data;
+
+#define COND_SEND(X) if (send(socket, X, strlen(X), 0) == -1) \
+    { \
+        Log(LOG_LEVEL_ERR, "Failed to write SMTP directive in %s:%i", __FILE__, __LINE__); \
+        return NULL; \
+    } \
+    else \
+    { \
+        fwrite(X, strlen(X), 1, stdout); \
+    }
+
+#define COND_RECV(X) if ((rcvd = recv(socket, X, sizeof(X), 0)) == -1) \
+    { \
+        Log(LOG_LEVEL_ERR, "Failed to read SMTP response in %s:%i", __FILE__, __LINE__); \
+        return NULL; \
+    } \
+    else \
+    { \
+        fwrite(X, rcvd, 1, stdout); \
+    }
+
+    char *sendbuf;
+    char recvbuf[CF_BUFSIZE];
+    int rcvd;
+    sendbuf = "220 test.com\r\n";
+    COND_SEND(sendbuf);
+    COND_RECV(recvbuf);
+    sendbuf = "250 Hello test.com, pleased to meet you\r\n";
+    COND_SEND(sendbuf);
+    COND_RECV(recvbuf);
+    sendbuf = "250 from@test.com... Sender ok\r\n";
+    COND_SEND(sendbuf);
+    COND_RECV(recvbuf);
+    sendbuf = "250 to@test.com... Recipient ok\r\n";
+    COND_SEND(sendbuf);
+    COND_RECV(recvbuf);
+    sendbuf = "354 Enter mail, end with \".\" on a line by itself\r\n";
+    COND_SEND(sendbuf);
+    while (true)
+    {
+        COND_RECV(recvbuf);
+        if ((rcvd == 3 && memcmp(recvbuf + rcvd - 3, ".\r\n", 3) == 0) ||
+            (rcvd > 3 && memcmp(recvbuf + rcvd - 4, "\n.\r\n", 4) == 0))
+        {
+            break;
+        }
+    }
+    sendbuf = "250 Message accepted for delivery\r\n";
+    COND_SEND(sendbuf);
+    COND_RECV(recvbuf);
+    sendbuf = "221 test.com closing connection\r\n";
+    COND_SEND(sendbuf);
+
+#undef COND_SEND
+#undef COND_RECV
+
+    cf_closesocket(socket);
+    free(data);
+
+    return NULL;
+}
+
+int ConnectToSmtpSocket(ARG_UNUSED const ExecConfig *config)
+{
+    int sockets[2];
+    if (socketpair(AF_UNIX, SOCK_STREAM, 0, sockets) < 0)
+    {
+        return -1;
+    }
+
+    int *thread_socket = xmalloc(sizeof(int));
+    *thread_socket = sockets[1];
```

Comment:
We cannot pass `sockets + 1` or `&sockets[1]` because the variable is stack allocated, and this stack frame gets destroyed after we have returned from this function. We also cannot pass the int cast as a pointer because `sizeof(int) != sizeof(int*)` on 64-bit platforms. They are 4 and 8, respectively.